business/offering data models and helper scripts

anizeani opened the pull request:
closes #171
closes #172
Instead of passing the json file path in the terminal, i open a dialog where the user selects the file.. I thought it to be more user friendly.

anizeani committed:
added business jsons

added business jsons

anizeani committed:
add product jsons

brenzi commented on client/test-data/bazaar/technopark/businesses/KücheEdison.json:
```diff
@@ -0,0 +1,7 @@
+{
+	"name": "Küche Edison",
+	"description": "wir sind ein lokales Geschäft im Technopark und offerieren köstlichen Kaffe",
```
this will be visible worldwide. I'd keep it informal "bei uns gibts köstlichen Kaffee". Shouldn't sound serious in any way

brenzi commented on client/test-data/bazaar/technopark/offerings/EdisonKaffee.json:
```diff
@@ -0,0 +1,4 @@
+{
```
on folder naming:

please replace "technopark" with "EdisonPaula"
subfolders "businesses / offerings do not make sense for me as it isn't clear if you want to group by business and its offerings, or as you did. I suggest to reflect `business` and `product` in filenames to reflect the type

anizeani replied:
i now created a folder structure of test-data/bazaar/business/products, ok?

anizeani committed:
clean up folder structure, clean up description in businesses

anizeani committed:
wip

anizeani committed:
uplaod single files to ipfs and return cid in terminal

anizeani committed:
register business by selecting json file, i think its more user friendly than pasting the whole path in the terminal

anizeani committed:
improve helper text

anizeani committed:
change folder naming

anizeani committed:
updated jsons

brenzi commented on client/test-data/bazaar/EdisonPaula/businesses/küchepauli/KüchePauli.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "Küche Pauli",
-	"description": "wir sind ein lokales Geschäft im Technopark und offerieren genussreichen Kaffe",
+	"description": "bei uns giebt es genussreichen Kaffe",
```
giebt

anizeani committed:
make community identifier and bizaccount required, comment out account creation, expect already existing account

anizeani committed:
make community identifier and bizaccount required, comment out account creation, expect already existing account
uploading product.json to ipfs
verifying that name and description are present in json file
add warning if logo is missing
create offering and upload to ipfs

anizeani committed:
removing optional account creation

anizeani committed:
removing optional account creation

anizeani committed:
typo

anizeani committed:
remove comment

brenzi commented on client/test-data/bazaar/EdisonPaula/businesses/kücheedison/KücheEdison.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "Kueche Edison",
```
use Umlauts plz. this has to work

anizeani replied:
sure, i'm on that. works for businesses, need to adapt for offerings :)

brenzi commented on client/test-data/bazaar/EdisonPaula/businesses/kücheedison/KücheEdison.json:
```diff
@@ -0,0 +1,7 @@
+  "longitude": "8.515962660312653",
+  "latitude": "47.390349148891545",
+  "logo": "QmesWqTQ7Hp7o6wCebvNiXUAuwRvGJ17ZDe8viYPWufvDH"
+}
```
As this is an example, please use all the fields available that are defined in
https://github.com/encointer/pallets/blob/master/bazaar/README.md

so, plz add:
* category: food
* address: Technoparkstr. 1, 8005 Zürich
* telephone: None
* email: None
* photos: <ipfs cid to json which is a folder with all pictures?? or rather a json listing all photos with a label? Do what is easier>
* openingHours: please check the Bazaar app data model: [here](https://github.com/encointer/encointer-wallet-flutter/blob/510e88f83c19a49c49dd2f5fade5f78a28851f50/lib/page-encointer/bazaar/shared/data_model/model/bazaarItemData.dart#L60) and [here](https://github.com/encointer/encointer-wallet-flutter/blob/510e88f83c19a49c49dd2f5fade5f78a28851f50/lib/page-encointer/bazaar/shared/data_model/model/bazaarItemData.dart#L103) and do Mon-Fri 8h-18h

same for other businesses

brenzi commented on client/test-data/bazaar/EdisonPaula/businesses/kücheedison/products/EdisonKaffee.json:
```diff
@@ -0,0 +1,5 @@
+{
+	"name": "Kaffee",
+	"description": "Koestlicher Kaffe",
+	"logo": "QmZzkgNe6B6M9Y3UeGgugwEB56v5qBm35bcPTPnZFNtY7d"
```
please read the spec carefully! this should say `image`, not `logo`

brenzi commented on client/test-data/bazaar/EdisonPaula/businesses/kücheedison/products/EdisonKaffee.json:
```diff
@@ -0,0 +1,5 @@
+	"name": "Kaffee",
+	"description": "Koestlicher Kaffe",
+	"logo": "QmZzkgNe6B6M9Y3UeGgugwEB56v5qBm35bcPTPnZFNtY7d"
+}
```
please add itemCondition: None (or new`?)

anizeani replied:
what about Identifier? 
and category? (again food?)

brenzi commented on client/test-data/bazaar/EdisonPaula/businesses/küchepauli/products/PauliKaffee.json:
```diff
@@ -0,0 +1,5 @@
+{
+	"name": "Kaffee",
+	"description": "Leckerer Kaffe",
+	"logo": "QmUhCgPrqep8huFDDjGdVneMfqKYxBFPyFvd3CjfwNYTPb"
```
dito

brenzi commented on client/upload-image-to-ipfs-and-return-cid.py:
```diff
@@ -0,0 +1,38 @@
+from tkinter import filedialog
+import os
+
+BUSINESSES_PATH = './test-data/bazaar/'
```
unused. delete!

brenzi commented on client/upload-image-to-ipfs-and-return-cid.py:
```diff
@@ -0,0 +1,38 @@
+        except:
+            print("failed to add image to ipfs")
+
+    print("cid of your file:", image_cid)
```
would be much more usable (i.e. in bash scripts) if you just returned the cid without any other printouts to stdout

brenzi commented on client/upload-image-to-ipfs-and-return-cid.py:
```diff
@@ -0,0 +1,38 @@
+        try:
+            image_cid = Ipfs.add(logo_path, ipfs_local)
+        except:
+            print("failed to add image to ipfs")
```
please exit with an error code and don't just print an error.

brenzi replied:
see client.py for examples how to do that

anizeani replied:
I've seen how you check the return codes and depending on them raise a specific error. 
in our case we do ipfs.add, which either adds remotely or locally. 
In any case we are in an elevated level, so the error would emerge from deeper in code. 
So for the curl command (remote ipfs add) there are 96 error return codes. (https://everything.curl.dev/usingcurl/returns) 
any of them are actually already handled below, we just need to add the "check" option where the curl command gets excecuted (see my latest file change in file py_client/ipfs.py.
I'm not sure I understand if you want me to handle those errors in the ipfs class by myself, because those errors are actually already handled by the subprocess.run function https://github.com/python/cpython/blob/main/Lib/subprocess.py#L528
what i did now is, enable the error raising in the deeper level and i extend the error output in case of raised error

brenzi replied:
just define a positive integer number which isn't used already and throw that. Don't care what the exact error was deeper down. just ensure we exit with a positice exit code

brenzi reviewed (changes requested):
please never use umlauts in file/folder names
On the other hand: I would use Umlauts in the Description because that will be a Test we want to pass

anizeani committed:
remove unused imports, remove prints

anizeani committed:
adding script to upload folders and returns cid

anizeani committed:
remove unused imports/path variable,
update documentation for script

anizeani committed:
updated json files, removed images from git

anizeani committed:
change none to null since none is only python type and not in the json format

anizeani committed:
updated None to null

anizeani committed:
added silent -s option to not display progress bar in terminal when uploading to remote ipfs node

anizeani committed:
silence progress output but show real errors

anizeani committed:
updating helper to only display cid and not print empty line (which i don't get yet why it's there)

anizeani committed:
add check=True to actually raise error if returncode is != 0

anizeani committed:
let error propagate up and return extended error message

brenzi commented:
Now we have quite a mess of scripts:
* register-business-simple (now broken because of path change)
* register-businesses (now broken because of path change)
* register-business
* register-offering
* upload-folder-to-ipfs-and-return-cid (sounds generic, but in the help comment it says `Register a business on chain`)
* upload-folder-to-ipfs-and-return-cid

then we also have 
* publish-icons.py which also just uploads stuff to ipfs

And there is no piece of doc in the README about what to use when and how. 

I'll need more time to review and find a solution. Please stop working on this PR for now @anizeani .

anizeani committed:
update documentation

anizeani committed:
refactoring: now using bazaar.py to enable register business or register offering

anizeani committed:
refactoring: now using bazaar.py to enable register business or register offering

anizeani committed:
remove the two scripts, now one script should do it all

anizeani committed:
remove price as argument

brenzi committed:
Merge branch 'master' into implement_business_and_offering_data_models_and_helper_scripts

brenzi committed:
cleanup and docs

brenzi committed:
let CI fail if registering biz or offerings fails

brenzi reviewed (approved):
cleaned up bits and pieces, streamlined and documented. all works locally
